mpa/modules/models/heads/ib_loss_head.py

```diff
@@ -0,0 +1,31 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from mmcls.models.builder import HEADS
+from mmcls.models.heads import LinearClsHead
+from mpa.modules.models.heads.non_linear_cls_head import NonLinearClsHead
+
+
+@HEADS.register_module()
+class IBLossHead(LinearClsHead):
```

Comment:
I don't understand that the head is added for the new loss.
I don't think this is the right way to do it if it's just to get the current epoch (loss per epoch). What do you think?
Does this have any other intentions?

Reply:
Because IB loss requires to have features before the last layer, not only the final output.

Reply:
It seems that you can cover it with a hook (example -> hook for get feature before the last layer) without a separate head. The reason why I am concerned about adding a head is that it feels strange in structure to change the head for each loss.

Reply:
I don't get it. Is there a way to reach to the feature of last layer by hook? I thought hook is only involved in the after/before training, not in the middle of training iteration. Do you suggest to make other custom EpochBasedRunner to reach to the last layer?

Reply:
In fact, I didn't really understand how to use loss. But what I want to say is that it is not a good logic to add a new head according to loss, I agree with Vlad's comment above.

Reply:
I'm not sure for detailed IB loss implementation, but if the thing that you want is extracting feature vector from last layer, registering additional forward hook before model update can be help for you. Of course, forward hook will require overhead, but clear code might have more merits for us.

Reply:
What Jihwan suggests is a little inefficient because it's like doing forward passes twice, causing large overhead. Getting feature vector before the last layer for IB loss is inevitable. There are two choice for this implementation. (If curious about detailed implementation, refer to [this page](https://github.com/pseulki/IB-Loss/blob/b29b8b005f614292e9ae8b43903a6f7df8ef09a6/models/resnet_cifar.py#L323) or original paper) 1) Making IB loss classifier, 2) Making IB loss head. Since we are using SAMClassifier, 1) need to inherit SAMClassifier, which is more hardcoded method. 2) is the better choice, I think. Reflecting comments of @harimkang and @sovrasov, I will try to implicitly implement the head for IB loss by renaming or wrapping it as custom head. If have any comment, please let me know.